[FIX] Make project_functional_block installable

tungtt opened the pull request:
Fix inheritance of search view of project task.

Commit:
[FIX] Make project_functional_block installable

Commit:
[FIX] Make project_functional_block installable

Commit:
Merge branch '9.0-functional-block' of https://github.com/tungtt/project into 9.0-functional-block

Commit:
[FIX] Fix inheritance error on form view of task because field progress on task form is now moved to module project_timesheet

dreispt commented:
Thank you @tungtt , TravisCI testes are not running properly yet on 9.0 unfortunately.
But in the meanwhile, could you please also fix the style checks?

tungtt commented:
My purpose was just to make it installable. However, results produced by TravisCI look cool so I will fix the styles.

@dreispt how can I assign this pull request to me so that when everything is OK, I will reassign it to you?

And sorry for this messy pull request (this is my first pull request in OCA).

Commit:
[FIX] Correct mistyped field name

dreispt commented:
Nothing messy, it's all good.
No need for the reassignments. Just comment here when it's ready for reviewers.

dreispt commented:
@tungtt OK, 9.0 tests are now fixed.
However we are asking for v9 modules to be in the new API, that's one of the reason why pylint checks are [failing](https://travis-ci.org/OCA/project/jobs/89060660#L258).
Could you please convert to the new API and check the remaining style errors?

dreispt commented:
PS: please post here if you need help or advice.

Commit:
[REF][IMP] Migrate module project_functional_block to new API. Removing unnecessary fields. Improve views

Commit:
[REF] Remove vim comment

tungtt commented:
@dreispt: It is ready to review, finally. This is a summary of my changes
- Convert to new API.
- Update directory structure to make the code cleaner. Each file contains only one model.
- Remove field `sequence` as I think we don't need it. Ordering functional blocks by complete names is enough.
- Add smart buttons to display number of child blocks and number of tasks.
- Refactoring the implementation of field `complete_name` to make it faster.
  - Functional blocks data are kind of stable so I think it is not worthy to do the computation of `complete_name` every time we read a functional block. Therefore, I removed the implementation in `name_get`.
  - I make it a normal field and handle the value update within `create` and `write` functions. I tried to make it a functional field but failed with the recursion update.
    - Tried to depend on `parent_id.complete_name` but failed because the update of functional field does not go through `write` function.
    - Tried to put a function in `@api.multi(_get_related_fblocks)` to modify the value of `self` passed in the compute function but failed. The code looks like this

```
@api.multi
def _get_related_fblocks(self):
    # code here, returns browse records

@api.multi(_get_related_fblocks)
@api.depends('name', 'parent_id.name')
def _compute_complete_name(self):
    # code here

complete_name = fields.Char(compute='_compute_complete_name', store=True, string='Complete Name', index=True, size=512)
```

```
- Tried to do this but failed
```

```
@api.multi
@api.depends('name', 'parent_id.name')
def _compute_complete_name(self):
    # search for fblocks which are in self and children of self
    # update complete names of all above records

complete_name = fields.Char(compute='_compute_complete_name', store=True, string='Complete Name', index=True, size=512)
```

However, current implementation of mine leads to an issue that `complete_name` is not translated when `name` is translated. Not sure how to fix this properly.

Regards,

P.S.: what is the keyword to replace `description` in `__openerp__.py` ?

dreispt commented:
> P.S.: what is the keyword to replace description in **openerp**.py ?

The README.rst file replaces the description key.

Commit:
[REF] Replace the description in __openerp__ by README file

Commit:
[REF] Remove vim comment

Commit:
[REF] Refactor directory structure

jgrandguillaume commented:
Hi @tungtt 

Thanks for the good work and congrats for your first PR in the OCA world. One last thing : you can take this model here for the Readme.rst https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

More generally, you can find an OCA official template module here (useful to get some best practice): https://github.com/OCA/maintainer-tools/tree/master/template/module

I'm giving you my  :+1:  once the Readme.rst is ok

Regards,

Joël

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,152 @@
+            ctx = self._context.copy()
+            ctx.update({'updating_complete_name': True})
+            self.with_context(ctx)._update_complete_name()
+        return res
```
I feel that the logic in create() and writ()e could be replaced by a converting complete_name into a computed field.
But I won't pick on that, just add it to the Known Issues/Roadmap section of the README please.

Commit:
[IMP] Change README.md to README.rst of project_functional_block

tungtt commented:
Done

dreispt commented:
@tungtt You still need to address Joël's comment above to "collect" his +1.

Commit:
[IMP] Update README.rst to match standards of OCA

Commit:
[REF] Correct W503 warnings

======== Testing test_flake8 ========
project_functional_block/models/project_functional_block.py:114:13: W503 line break before binary operator
project_functional_block/models/project_functional_block.py:151:13: W503 line break before binary operator

tungtt commented:
@jgrandguillaume, @dreispt `README.rst` is updated.

I also correct W503 warnings. However, I see TravisCI says this

```
======== Testing test_pylint ========
From https://github.com/OCA/project
 * [new branch]      9.0        -> origin/9.0
warning: refname 'origin/9.0' is ambiguous.
Start lint check just in modules changed
************* Module project_functional_block.__openerp__
project_functional_block/__openerp__.py:22: [C8102(manifest-required-key), ] Missing required key "images" in manifest file
```

What is the key `images` for ?

dreispt commented on project_functional_block/__openerp__.py:
```diff
@@ -19,25 +19,28 @@
- 'application': True,
+{
+    'name': 'Project requiring functional blocks',
+    'version': '1.2',
```
Set to `9.1.1.2.0`

tungtt replied:
done.

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,152 @@
+
+    # Columns
+    name = fields.Char('Name', size=64, required=True, translate=True)
+    code = fields.Char('Code', size=8)
```
The current best practice is to remove `size` from Char fields

tungtt replied:
I am not sure that it is the best practice. For me, and also in my company, the best practice is to specifying explicitly the size of each `Char` field so that
- We have a clear understanding how much characters we can input in a field
- We can reduce the size of the database.

I will wait for your feedback to do this modification.

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,152 @@
+    name = fields.Char('Name', size=64, required=True, translate=True)
+    code = fields.Char('Code', size=8)
+    # TODO: this field doesn't work with multi-languages
+    complete_name = fields.Char(string="Complete Name", index=True)
```
Add a TODO comment: to convert into a computed field

dreispt commented on project_functional_block/views/project_functional_block_view.xml:
```diff
@@ -0,0 +1,83 @@
+                    <field name="complete_name" />
+                    <field name="parent_id" />
+                    <group expand="0" string="Group By...">
+                        <filter string="Parent" icon="terp-partner"
```
Here `icon` is deprecated, and I believe that `expand`also.

tungtt replied:
done

dreispt commented on project_functional_block/views/project_functional_block_view.xml:
```diff
@@ -0,0 +1,83 @@
+                            <label class="oe_edit_only" for="name" string="Name"/>
+                            <h1><field name="name" placeholder="Name"/></h1>
+                        </div>
+                        <group col="4">
```
`col`is deprecated

tungtt replied:
done. But I find it inconvenient because I need to add extra `group`s.

dreispt commented on project_functional_block/views/project_task_view.xml:
```diff
@@ -0,0 +1,48 @@
+        <record id="view_task_functional_block_form" model="ir.ui.view">
+            <field name="name">project.task.functional_block.form</field>
+            <field name="model">project.task</field>
+            <field name="type">form</field>
```
Views `type`attribute is deprecated - remove form other views also.

tungtt replied:
done

dreispt commented on project_functional_block/views/project_task_view.xml:
```diff
@@ -0,0 +1,48 @@
+                	<separator orientation="vertical"/>
+                    <filter string="Functional Block"
+                        name="groupby_functional_block_id"
+                        icon="terp-gtk-jump-to-ltr"
```
remove icon

tungtt replied:
done

dreispt commented:
Well done @tungtt.
Some minor fixes a it's ready to go.
PS: if you can make `complete_name` a computed field that would be great!

dreispt commented:
@tungtt Ping!

Commit:
resolve conflict after pull tungtt/project:9.0-functional-block subtree

Commit:
[REF] Remove deprecated keywords. Convert complete_name to function field

Commit:
[REF] Remove print keyword

dreispt commented on project_functional_block/__openerp__.py:
```diff
@@ -19,25 +19,28 @@
- 'application': True,
+{
+    'name': 'Project requiring functional blocks',
+    'version': '9.1.1.2.0',
```
Version should begin with the Odoo target version,  9.0

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,146 @@
+    _parent_order = 'name'
+
+    # Columns
+    name = fields.Char('Name', size=64, required=True, translate=True)
```
Remove `size`from all Char fields

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,146 @@
+        string='Child Blocks Count', compute='_count_child_fblocks')
+
+    @api.multi
+    def _compute_complete_name(self):
```
I think this can be implemented with a single line:

```
fir fblock in self:
    fblock.complete_name = '%s / %s' % (fblock.parent_id.complete_name, fblock.name)
```

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,146 @@
+            func_block.complete_name = complete_name
+
+    @api.multi
+    def _count_tasks(self):
```
If you add a `child_ids` you can then:

    for fblock in self:
        fblock.tasks_count = sum(x.tasks_count fro x in fblock.child_ids)

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,146 @@
+            'target': action.target,
+            'context': str(ret_ctx),
+            'res_model': action.res_model,
+            'domain': "[('id', 'in', (" + task_ids_str + ",))]"
```
I don't think that domain and context must be strings.
AFAIK Python list and dict can be used directly..

dreispt commented on project_functional_block/models/project_functional_block.py:
```diff
@@ -0,0 +1,146 @@
+        }
+
+    @api.multi
+    def _count_child_fblocks(self):
```
If you add a `child_ids` you can then:

```
for fblock in self:
    fblock.child_fblocks_count = sum(x.child_fblocks_count fro x in fblock.child_ids)
```

dreispt commented:
I repeated the review. A few fixes and improvement suggestions made.
